File: app/schemas/pool.py
from datetime import datetime
from typing import List, Optional
from pydantic import BaseModel, ConfigDict


class PoolMemberResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)
    etf_code: str
    etf_name: Optional[str] = None
    added_at: Optional[datetime] = None
    notes: Optional[str] = None


class PoolBase(BaseModel):
    name: str
    description: Optional[str] = None


class PoolCreate(PoolBase):
    pass


class PoolUpdate(BaseModel):
    name: Optional[str] = None
    description: Optional[str] = None


class PoolResponse(PoolBase):
    model_config = ConfigDict(from_attributes=True)
    id: int
    members: List[PoolMemberResponse] = []
    created_at: Optional[datetime] = None
    updated_at: Optional[datetime] = None


class PoolMemberCreate(BaseModel):
    etf_code: str
    notes: Optional[str] = None

File: web/src/types/pool.ts
export interface Pool {
  id: number;
  name: string;
  description?: string;
  created_at: string;
  updated_at: string;
  members: PoolMember[];
}

export interface PoolMember {
  pool_id: number;
  etf_code: string;
  etf_name?: string;
  added_at?: string;
  is_active?: boolean;
}

export interface PoolWeight {
  pool_id: number;
  etf_code: string;
  etf_name?: string;
  target_weight: number;
  suggested_weight?: number;
  weight_source: 'manual' | 'equal' | 'score' | 'risk_parity';
}

export interface PoolAnalytics {
  pool_id: number;
  name: string;
  members: PoolWeight[];
  category_distribution: Record<string, { count: number; weight: number }>;
  correlation_matrix?: {
    codes: string[];
    matrix: number[][];
  };
  performance: {
    return_1m?: number;
    return_3m?: number;
    volatility_20d?: number;
    sharpe_1y?: number;
    max_drawdown?: number;
  };
  rebalance_needed?: Array<{
    code: string;
    target: number;
    actual: number;
    diff: number;
  }>;
}

export interface PoolSnapshot {
  id: number;
  pool_id: number;
  snapshot_date: string;
  data: Record<string, unknown>;
  created_at: string;
}

export interface PoolCreate {
  name: string;
  description?: string;
}

export interface PoolUpdate {
  name?: string;
  description?: string;
}
